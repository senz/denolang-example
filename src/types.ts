export interface Route {
	path: string;
	method: string;
	handler: HttpHandler;
	noauth?: boolean;
}

export interface Employees {
	[key: string]: string;
}

export interface EmployeeHierarchy {
	[key: string]: EmployeeHierarchy;
}

export type HttpHandler = (_request: Request) => Promise<Response>;
