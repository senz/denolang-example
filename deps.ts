export * as http from 'std/http/mod.ts';
export * as dotenv from 'https://deno.land/x/dotenv@v3.2.2/mod.ts';
export { default as Ajv } from 'https://cdn.skypack.dev/ajv?dts';
export * as ajv from 'https://cdn.skypack.dev/ajv?dts';
export * as mysql from 'https://deno.land/x/mysql@v2.11.0/mod.ts';
export { scryptBasicAuth } from 'https://deno.land/x/scrypt_basic_auth@1.0.2/mod.ts';
