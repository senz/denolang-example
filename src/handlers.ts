import { EmployeeHierarchy, Employees } from 'src/types.ts';
import { mysql } from '../deps.ts';

const buildHierarchy = (relations: Employees): EmployeeHierarchy => {
	let hierarchy: EmployeeHierarchy = {};
	const subordinates: EmployeeHierarchy = {};

	for (const [employee, supervisor] of Object.entries(relations)) {
		if (!subordinates[employee]) {
			subordinates[employee] = {};
		}

		if (subordinates[supervisor]) {
			subordinates[supervisor][employee] = subordinates[employee];
		} else {
			subordinates[supervisor] = { [employee]: subordinates[employee] };
		}
	}

	for (const [_employee, supervisor] of Object.entries(relations)) {
		if (!relations[supervisor]) {
			hierarchy = { [supervisor]: subordinates[supervisor] };
			break;
		}
	}

	return hierarchy;
};

export const handlerEmpBuilder =
	(validate: (data: Employees) => boolean, client: mysql.Client) =>
	async (_request: Request): Promise<Response> => {
		const contentType = _request.headers.get('content-type');
		if (contentType?.toLocaleLowerCase() !== 'application/json') {
			return Promise.resolve(
				new Response('"bad request: contenttype"', {
					status: 400,
					headers: {
						'content-type': 'application/json',
					},
				}),
			);
		}

		let body: Employees | undefined;
		try {
			body = await JSON.parse(await _request.text()) as Employees;
		} catch (_e) {
			return Promise.resolve(
				new Response('"bad request: invalid json"', {
					status: 400,
					headers: {
						'content-type': 'application/json',
					},
				}),
			);
		}
		console.log(body);

		if (!validate(body)) {
			return Promise.resolve(
				new Response('"bad request: validation error"', {
					status: 400,
					headers: {
						'content-type': 'application/json',
					},
				}),
			);
		}

		const result = buildHierarchy(body);
		console.log(result);

		await client.execute('truncate table employees;');

		let id = 1;
		const nameToId: { [key: string]: number } = {};

		// Function to flatten the hierarchy and insert it into the database
		async function insertHierarchy(
			supervisorId: number | null,
			hierarchy: EmployeeHierarchy,
		) {
			for (const name in hierarchy) {
				const _res = await client.execute(
					`INSERT INTO employees(id, name, supervisor_id) values(?, ?, ?)`,
					[id, name, supervisorId],
				);
				nameToId[name] = id;
				id += 1;
				await insertHierarchy(nameToId[name], hierarchy[name]);
			}
		}

		await insertHierarchy(null, result);

		return Promise.resolve(
			new Response(JSON.stringify(result), {
				status: 201,
				headers: {
					'content-type': 'application/json',
				},
			}),
		);
	};

export const handlerSvBuilder =
	(client: mysql.Client) => async (_request: Request): Promise<Response> => {
		const match = new URLPattern({ pathname: '/supervisors/:name/:levels' })
			.exec(_request.url);
		const groups = match?.pathname.groups;

		const res = await client.query(
			`    
  WITH RECURSIVE supervisor_cte AS (
      SELECT id, name, supervisor_id, 1 lvl
      FROM employees
      WHERE name = ?
      UNION ALL
      SELECT e.id, e.name, e.supervisor_id, lvl + 1
      FROM employees e
      INNER JOIN supervisor_cte scte ON e.id = scte.supervisor_id
  )
  SELECT name FROM supervisor_cte WHERE lvl BETWEEN 2 AND ? + 1;
  `,
			[groups?.name, groups?.levels],
		);

		console.log(res);

		return Promise.resolve(
			new Response(JSON.stringify(res), {
				status: 200,
				headers: {
					'content-type': 'application/json',
				},
			}),
		);
	};

export const okHandler = (_request: Request): Promise<Response> => {
	return Promise.resolve(
		new Response(JSON.stringify('ok'), {
			status: 200,
			headers: {
				'content-type': 'application/json',
			},
		}),
	);
};
