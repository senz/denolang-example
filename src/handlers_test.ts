import { mysql } from '../deps.ts';
import { mock, snapshot } from '../test.deps.ts';
import { handlerEmpBuilder, handlerSvBuilder } from 'src/handlers.ts';
import { Employees } from './types.ts';

class MockClient extends mysql.Client {}

Deno.test('emp: invalid content type. 400', async function (t): Promise<void> {
	const r = new Request('http://localhost/employees', {
		body: '',
		method: 'POST',
		headers: {
			'content-type': 'ololopishpish',
		},
	});

	const client = new MockClient();

	const validate = (_data: Employees): boolean => {
		return true;
	};

	const res = await handlerEmpBuilder(validate, client)(r);
	const ress = {
		headers: res.headers,
		body: await res.text(),
		status: res.status,
	};

	snapshot.assertSnapshot(t, ress);
});

Deno.test('emp: invalid json. 400', async function (t): Promise<void> {
	const r = new Request('http://localhost/employees', {
		body: 'sdfds/dfdfds',
		method: 'POST',
		headers: {
			'content-type': 'application/json',
		},
	});

	const client = new MockClient();

	const validate = (_data: Employees): boolean => {
		return true;
	};

	const res = await handlerEmpBuilder(validate, client)(r);
	const ress = {
		headers: res.headers,
		body: await res.text(),
		status: res.status,
	};
	snapshot.assertSnapshot(t, ress);
});

Deno.test('emp: malformed json. 400', async function (t): Promise<void> {
	const r = new Request('http://localhost/employees', {
		body: '{}',
		method: 'POST',
		headers: {
			'content-type': 'application/json',
		},
	});

	const client = new MockClient();

	const validate = (_data: Employees): boolean => {
		return false;
	};

	const res = await handlerEmpBuilder(validate, client)(r);
	const ress = {
		headers: res.headers,
		body: await res.text(),
		status: res.status,
	};
	snapshot.assertSnapshot(t, ress);
});

Deno.test(
	'emp: valid json. stored in db. 201',
	async function (t): Promise<void> {
		const body = `
{
    "Pete": "Nick",
    "Barbara": "Nick",
    "Nick": "Sophie",
    "Sophie": "Jonas"
}
`;
		const r = new Request('http://localhost/employees', {
			body,
			method: 'POST',
			headers: {
				'content-type': 'application/json',
			},
		});

		const client = new MockClient();
		const s = mock.stub(
			client,
			'execute',
			mock.resolvesNext([
				{ affectedRows: 5 }, // truncate
				{ affectedRows: 1 }, // inserts...
				{ affectedRows: 1 },
				{ affectedRows: 1 },
				{ affectedRows: 1 },
				{ affectedRows: 1 },
			]),
		);

		const validate = (_data: Employees): boolean => true;

		let res;
		try {
			res = await handlerEmpBuilder(validate, client)(r);
		} finally {
			s.restore();
		}

		const ress = {
			headers: res.headers,
			body: await res.text(),
			status: res.status,
		};
		snapshot.assertSnapshot(t, ress);
	},
);

Deno.test(
	'sv: 2 entries in db. array with elements returned. 200',
	async function (t): Promise<void> {
		const r = new Request('http://localhost/supervisors/Nick/2', {
			method: 'GET',
			headers: {},
		});

		const client = new MockClient();
		const s = mock.stub(
			client,
			'query',
			mock.resolvesNext([
				[
					{
						'name': 'Sophie',
					},
					{
						'name': 'Jonas',
					},
				],
			]),
		);

		let res;
		try {
			res = await handlerSvBuilder(client)(r);
		} finally {
			s.restore();
		}

		const ress = {
			headers: res.headers,
			body: await res.text(),
			status: res.status,
		};
		snapshot.assertSnapshot(t, ress);
	},
);
