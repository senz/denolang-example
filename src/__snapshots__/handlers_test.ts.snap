export const snapshot = {};

snapshot[`emp: invalid content type. 400 1`] = `
{
  body: '"bad request: contenttype"',
  headers: Headers { "content-type": "application/json" },
  status: 400,
}
`;

snapshot[`emp: invalid json. 400 1`] = `
{
  body: '"bad request: invalid json"',
  headers: Headers { "content-type": "application/json" },
  status: 400,
}
`;

snapshot[`emp: malformed json. 400 1`] = `
{
  body: '"bad request: validation error"',
  headers: Headers { "content-type": "application/json" },
  status: 400,
}
`;

snapshot[`emp: valid json. stored in db. 201 1`] = `
{
  body: '{"Jonas":{"Sophie":{"Nick":{"Pete":{},"Barbara":{}}}}}',
  headers: Headers { "content-type": "application/json" },
  status: 201,
}
`;

snapshot[`sv: 2 entries in db. array with elements returned. 200 1`] = `
{
  body: '["Sophie","Jonas"]',
  headers: Headers { "content-type": "application/json" },
  status: 200,
}
`;
